httpserver/reachability_testing.h

```diff
@@ -1,79 +1,129 @@
-class reach_record_t {
-
-  public:
-    // The time the node failed for the first time
-    // (and hasn't come back online)
-    time_point_t first_failure;
-    time_point_t last_failure;
-    // whether it's been reported to Oxend
-    bool reported = false;
-
-    // whether reachable over http
-    bool http_ok = true;
-    // whether reachable over zmq
-    bool zmq_ok = true;
-
-    reach_record_t();
+// Returns std::greater on the std::get<N>(v)th element value.
+template <typename T, size_t N>
+struct nth_greater {
+    constexpr bool operator()(const T& lhs, const T& rhs) const {
+        return std::greater<std::tuple_element_t<N, T>>{}(
+                std::get<N>(lhs), std::get<N>(rhs));
+    }
 };
-} // namespace detail
 
-enum class ReachType { HTTP, ZMQ };
-
-enum class ReportType { GOOD, BAD };
+struct incoming_test_state {
+    time_point_t last_test{};
+    time_point_t last_whine{};
+    bool was_failing = false;
+};
 
-class reachability_records_t {
+} // namespace detail
 
-    // TODO: sn_records are heavy (3 strings), so how about we only store the
-    // pubkey?
+class Swarm;
 
-    // Nodes that failed the reachability test
-    // Note: I don't expect this list to be large, so
-    // `std::vector` is probably faster than `std::set` here
-    std::unordered_map<sn_pub_key_t, detail::reach_record_t> offline_nodes_;
+class reachability_testing {
+  public:
+    // How often we tick the timer to check whether we need to do any tests.
+    inline static constexpr auto TESTING_TIMER_INTERVAL = 50ms;
+
+    // Distribution for the seconds between node tests: we throw in some randomness to avoid
+    // potential clustering of tests.  (Note that there is some granularity here as the test timer
+    // only runs every TESTING_TIMER_INTERVAL).
+    inline static thread_local std::normal_distribution<double> TESTING_INTERVAL{10.0, 3.0};
+
+    // The linear backoff after each consecutive test failure before we re-test.  Specifically we
+    // schedule the next re-test for (TESTING_BACKOFF*previous_failures) + TESTING_INTERVAL(rng).
+    inline static constexpr auto TESTING_BACKOFF = 10s;
+
+    // The upper bound for the re-test interval.
+    inline static constexpr auto TESTING_BACKOFF_MAX = 2min;
+
+    // The maximum number of nodes that we will re-test at once (i.e. per TESTING_TIMING_INTERVAL);
+    // mainly intended to throttle ourselves if, for instance, our own connectivity loss makes us
+    // accumulate tons of nodes to test all at once.  (Despite the random intervals, this can happen
+    // if we also get decommissioned during which we can't test at all but still have lots of
+    // failing nodes we want to test right away when we get recommissioned).
+    inline static constexpr int MAX_RETESTS_PER_TICK = 4;
+
+    // Maximum time without a ping before we start whining about it.
+    //
+    // We have a probability of about 0.368 of *not* getting pinged within a second ping interval
```

Comment:
why is the probability `0.368`?

Reply:
It's (approximately) the value of
![image](https://user-images.githubusercontent.com/4459524/114740883-a5234b00-9d20-11eb-955b-bceef74a0a58.png)
but I thought it would be going too far to put that in a comment.

Reply:
it could help in the future to have a citation to this fact in that comment